### PR TITLE
add count observation for input DataObjects (#782)

### PR DESCRIPTION
**Pull request by zzeekk**
see #782

### CODE REVIEW

**Comment by zzeekk** [comment]
Implementation is nearly final from my point of view.
Open points 
- Tests for UniqueKeyExpectation scope = Job/All
- Tests for validate Expectations from Input only DataObjects
- Review of definition of Input only DataObjects: instead of taking all DataObjects which are never used as Output in the DAG, we probably should limit the DAG first to all DataObjects implementing ExpectationValidation (e.g. exclude FileTransferAction, and SparkFileDataObjects...)

File: sdl-core/src/main/scala/io/smartdatalake/config/InstanceRegistry.scala [hunk 1]
```diff
@@ -95,4 +105,22 @@ class InstanceRegistry {
    * Returns registered Connections
    */
   def getConnections: Seq[Connection] = instances.values.collect{ case c: Connection => c}.toSeq
+
+  /**
+   * Returns Ids of DataObjects that are used as input only, e.g. there is no Action having these DataObjects as output.
+   * Value is precomputed to avoid evaluation for every Action.
+   */
+  def getInputOnlyDataObjectIds: Seq[DataObjectId] = {
+    //TODO should we only consider DataObjects with ExpectationValidation?
```
Comment (Geheiner) [inline sdl-core/src/main/scala/io/smartdatalake/config/InstanceRegistry.scala]: You could keep this method and create a new one that then additionally filters for DataObjects with ExpectationValidation

File: sdl-core/src/main/scala/io/smartdatalake/workflow/dataframe/spark/SparkDataFrame.scala [hunk 2]
```diff
@@ -23,10 +23,12 @@ import io.smartdatalake.config.SdlConfigObject.DataObjectId
 import io.smartdatalake.definitions.Environment
 import io.smartdatalake.util.hdfs.PartitionValues
 import io.smartdatalake.util.misc.SchemaUtil
+import io.smartdatalake.util.spark.PushPredicateThroughTolerantCollectMetricsRuleObject.tolerantMetricsMarker
```
Comment (Geheiner) [inline sdl-core/src/main/scala/io/smartdatalake/workflow/dataframe/spark/SparkDataFrame.scala]: unused import

File: sdl-core/src/main/scala/io/smartdatalake/workflow/action/DataFrameActionImpl.scala [hunk 3]
```diff
@@ -80,6 +81,17 @@ abstract class DataFrameActionImpl extends ActionSubFeedsImpl[DataFrameSubFeed]
    */
   def saveModeOptions: Option[SaveModeOptions] = None
 
+  /**
+   * List of expectation definitions to evaluate when executing this Action, see [[Expectation]] for details.
+   *
+   * Note: Expectations defined DataObjects measure data quality and are evaluated against the output only.
```
Comment (Geheiner) [inline sdl-core/src/main/scala/io/smartdatalake/workflow/action/DataFrameActionImpl.scala]:
```suggestion
   * Note: Expectations defined on DataObjects measure data quality and are evaluated against the output only.
```

File: sdl-core/src/main/scala/io/smartdatalake/workflow/action/DataFrameActionImpl.scala [hunk 4]
```diff
@@ -249,11 +261,28 @@ abstract class DataFrameActionImpl extends ActionSubFeedsImpl[DataFrameSubFeed]
     if (breakDataFrameLineage || preparedSubFeed.isStreaming.contains(true) || preparedSubFeed.filter.isDefined) preparedSubFeed = preparedSubFeed.breakLineage
     // enrich with fresh DataFrame if needed
     preparedSubFeed = enrichSubFeedDataFrame(input, preparedSubFeed, context.phase, isRecursive)
+    // add observations on input DataFrame
+    if (Environment.enableInputDataObjectCount) {
+      input match {
+        case evDataObject: ExpectationValidation => preparedSubFeed.dataFrame.foreach { df =>
+          // collect additional aggregate expressions for this DataObject from Action expectations with scope Job
+          val inputJobAggExpressionColumns = expectations.filter(_.scope == ExpectationScope.Job).flatMap(_.getInputAggExpressionColumns(id))
+          val additionalJobAggExpressionColumns = inputJobAggExpressionColumns.filter(c => c.getName.exists(n => n.endsWith(s"#${input.id.id}"))).map(c => c.as(c.getName.get.stripSuffix(s"#${input.id.id}"))) ++
```
Comment (Geheiner) [inline sdl-core/src/main/scala/io/smartdatalake/workflow/action/DataFrameActionImpl.scala]: can you split this into 2 vals that you then concatenate in to become additionalJobAggExpressionColumns?
If I have 2 vals with a good name it's easier to read to see what you are actually extracting here.
I also have difficulty to understand where the counting is performed

File: sdl-core/src/main/scala/io/smartdatalake/workflow/action/DataFrameActionImpl.scala [hunk 5]
```diff
@@ -263,10 +292,30 @@ abstract class DataFrameActionImpl extends ActionSubFeedsImpl[DataFrameSubFeed]
     // apply expectation validation
     output match {
       case evDataObject: DataObject with ExpectationValidation =>
-        val (dfExpectations, observation) = evDataObject.setupConstraintsAndJobExpectations(subFeed.dataFrame.get)
+        // collect additional aggregate expressions for from Action expectations with scope Job
+        val additionalJobAggExpressionColumns = expectations.filter(_.scope == ExpectationScope.Job).flatMap(_.getAggExpressionColumns(evDataObject.id))
+        // setup observation
+        val (dfExpectations, outputObservation) = evDataObject.setupConstraintsAndJobExpectations(subFeed.dataFrame.get, additionalJobAggExpressionColumns = additionalJobAggExpressionColumns)
+        // combine input observations with output observation into one combined observation
```
Comment (Geheiner) [inline sdl-core/src/main/scala/io/smartdatalake/workflow/action/DataFrameActionImpl.scala]: can you explain in this comment why you add the input observation into the outputObservation ? From the context I assume it is related to the fact that you want to compare input and output count? Hard to follow without context

File: sdl-core/src/main/scala/io/smartdatalake/workflow/action/DataFrameActionImpl.scala [hunk 6]
```diff
@@ -263,10 +292,30 @@ abstract class DataFrameActionImpl extends ActionSubFeedsImpl[DataFrameSubFeed]
...
+        outputObservation match {
+          case outputSparkObservation: SparkObservation =>
+            inputSubFeeds.flatMap(_.observation).collect{ case x: SparkObservation => x}.map(_.getName)
+            val inputSparkObservationNames = inputSubFeeds.flatMap(_.observation).collect{ case x: SparkObservation => x}.map(_.getName)
+            outputSparkObservation.setOtherObservationNames(inputSparkObservationNames)
+        }
+        val inputObservationsToCombine = inputSubFeeds.flatMap { subFeed =>
+          subFeed.observation match {
+            case Some(_: SparkObservation) => None // ignore as this is handled by output observation
+            case Some(otherObservation) => Some(PrefixedObservation(otherObservation, subFeed.dataObjectId.id+"#")) // add input DataObjectId prefix to metrics
```
Comment (Geheiner) [inline sdl-core/src/main/scala/io/smartdatalake/workflow/action/DataFrameActionImpl.scala:317]: also not really sure what "combining obervations" means ans why we need to do it, another comment would be nice

File: sdl-core/src/main/scala/io/smartdatalake/workflow/DataFrameSubFeed.scala [hunk 7]
```diff
@@ -131,7 +131,7 @@ object DataFrameSubFeed {
     assert(elements.forall(_.subFeedType =:= expectedTpe), s"Unsupported subFeedType(s) ${elements.filter(c => !(c.subFeedType =:= expectedTpe)).map(_.subFeedType.typeSymbol.name).toSet.mkString(", ")} in method $parentMethod")
   }
 
-  private[smartdatalake] lazy val getKnownSubFeedTypes: Seq[Type] = {
+  @transient private[smartdatalake] lazy val getKnownSubFeedTypes: Seq[Type] = {
```
Comment (Geheiner) [inline sdl-core/src/main/scala/io/smartdatalake/workflow/DataFrameSubFeed.scala:134]: what is the tale behind this change ;)

File: sdl-core/src/main/scala/io/smartdatalake/workflow/action/DataFrameActionImpl.scala [hunk 8]
```diff
@@ -80,6 +81,17 @@ abstract class DataFrameActionImpl extends ActionSubFeedsImpl[DataFrameSubFeed]
    */
   def saveModeOptions: Option[SaveModeOptions] = None
 
+  /**
+   * List of expectation definitions to evaluate when executing this Action, see [[Expectation]] for details.
+   *
+   * Note: Expectations defined DataObjects measure data quality and are evaluated against the output only.
+   * Expectations defined on Actions measure quality of the transformation process and can measure and compare metrics between inputs and main output.
```
Comment (Geheiner) [inline sdl-core/src/main/scala/io/smartdatalake/workflow/action/DataFrameActionImpl.scala]:
```suggestion
   * Expectations defined on Actions measure quality of the transformation process and can measure and compare metrics between all input DataObjects and the main output DataObject.
```

File: sdl-core/src/main/scala/io/smartdatalake/workflow/action/DataFrameActionImpl.scala [hunk 9]
```diff
@@ -249,11 +261,28 @@ abstract class DataFrameActionImpl extends ActionSubFeedsImpl[DataFrameSubFeed]
...
+    // add observations on input DataFrame
+    if (Environment.enableInputDataObjectCount) {
+      input match {
+        case evDataObject: ExpectationValidation => preparedSubFeed.dataFrame.foreach { df =>
+          // collect additional aggregate expressions for this DataObject from Action expectations with scope Job
+          val inputJobAggExpressionColumns = expectations.filter(_.scope == ExpectationScope.Job).flatMap(_.getInputAggExpressionColumns(id))
+          val additionalJobAggExpressionColumns = inputJobAggExpressionColumns.filter(c => c.getName.exists(n => n.endsWith(s"#${input.id.id}"))).map(c => c.as(c.getName.get.stripSuffix(s"#${input.id.id}"))) ++
+            inputJobAggExpressionColumns.filter(c => c.getName.exists(n => !n.contains("#") && input.id == prioritizedMainInputCandidates.head.id))
+          // for global input, validate constraints and expectations, otherwise just add default expectations, e.g. count
+          val isGlobalInputOnly = context.instanceRegistry.isInputOnlyDataObject(subFeed.dataObjectId)
```
Comment (Geheiner) [inline sdl-core/src/main/scala/io/smartdatalake/workflow/action/DataFrameActionImpl.scala:273]: what do you mean with globalInput? Do you mean that it's the input of the whole Action and not like the input of the second transformation in case there are mutlitple transformations of the action?

Comment (Geheiner) [inline sdl-core/src/main/scala/io/smartdatalake/workflow/action/DataFrameActionImpl.scala]:
```suggestion
          val isInputOnlyDataObject = context.instanceRegistry.isInputOnlyDataObject(subFeed.dataObjectId)
```

what do you mean with globalInput? Do you mean that it's the input of the whole Action and not like the input of the second transformation in case there are mutlitple transformations of the action?

File: sdl-core/src/main/scala/io/smartdatalake/workflow/dataobject/ExpectationValidation.scala [hunk 10]
```diff
@@ -55,48 +59,41 @@ private[smartdatalake] trait ExpectationValidation { this: DataObject with Smart
    */
   def expectations: Seq[Expectation]
 
-  def setupConstraintsAndJobExpectations(df: GenericDataFrame)(implicit context: ActionPipelineContext): (GenericDataFrame, DataFrameObservation) = {
+  def setupConstraintsAndJobExpectations(df: GenericDataFrame, defaultExpectationsOnly: Boolean = false, predicateTolerant: Boolean = false, additionalJobAggExpressionColumns: Seq[GenericColumn] = Seq())(implicit context: ActionPipelineContext): (GenericDataFrame, DataFrameObservation) = {
     // add constraint validation column
-    val dfConstraints = setupConstraintsValidation(df)
+    val dfConstraints = if (defaultExpectationsOnly) df else setupConstraintsValidation(df)
```
Comment (Geheiner) [inline sdl-core/src/main/scala/io/smartdatalake/workflow/dataobject/ExpectationValidation.scala:64]: where do the default expectations come from?

Comment (Geheiner) [inline sdl-core/src/main/scala/io/smartdatalake/workflow/dataobject/ExpectationValidation.scala:75]: I dont understand why you need to make this if distinction here. When the user specifies no constraints, then contraints is empty and therfore setupConstraintsValidation does nothing anyway

File: sdl-core/src/main/scala/io/smartdatalake/workflow/dataobject/ExpectationValidation.scala [hunk 11]
```diff
@@ -55,48 +59,41 @@ private[smartdatalake] trait ExpectationValidation { this: DataObject with Smart
...
     // setup job expectations as DataFrame observation
     val jobExpectations = expectations.filter(_.scope == ExpectationScope.Job)
     val (dfJobExpectations, observation) = {
       implicit val functions: DataFrameFunctions = DataFrameSubFeed.getFunctions(df.subFeedType)
-      val expectationColumns = (defaultExpectations ++ jobExpectations).flatMap(_.getAggExpressionColumns(this.id))
-      setupObservation(dfConstraints, expectationColumns, context.isExecPhase)
+      val expectationColumns = (defaultExpectations ++ (if (defaultExpectationsOnly) Seq() else jobExpectations)).flatMap(_.getAggExpressionColumns(this.id)) ++ additionalJobAggExpressionColumns
```
Comment (Geheiner) [inline sdl-core/src/main/scala/io/smartdatalake/workflow/dataobject/ExpectationValidation.scala]: how can there be jobExpectations if the user defined no expectation? I would assume jobExpectations is empty anyway if the user did not specify any expectations

File: sdl-core/src/main/scala/io/smartdatalake/workflow/dataframe/DataFrameFunctions.scala [hunk 12]
```diff
@@ -31,6 +31,9 @@ import scala.reflect.runtime.universe
 trait DataFrameFunctions {
   protected def subFeedType: universe.Type
 
+  // Attention: Don't name this method getSubFeedType, Scala will otherwise compile it as a property and StatusInfoServer will try to serialize it and get an error "Direct self-reference leading to cycle..."
```
Comment (Geheiner) [inline sdl-core/src/main/scala/io/smartdatalake/workflow/dataframe/DataFrameFunctions.scala:34]: wow, what monster did I build :D

File: sdl-core/src/main/scala/io/smartdatalake/workflow/action/DataFrameActionImpl.scala [hunk 13]
```diff
@@ -278,11 +328,24 @@ abstract class DataFrameActionImpl extends ActionSubFeedsImpl[DataFrameSubFeed]
...
+            case sparkObservation: SparkObservation => sparkObservation.setOtherObservationsPrefix(id.id+"#")
+            case _ => ()
+          }
+          observation.waitFor()
+        }.getOrElse(Map())
+        // get input metrics with scope All
+        val inputMetrics = if (isMainOutput) getInputAggMetrics(subFeed) else Map()
+        // if this is mainOutput, enrich main input metrics
+        val enrichmentFunc: Map[String,_] => Map[String,_] = if (isMainOutput) enrichMainInputMetrics else identity
+        // evaluate and validate expectations
```
Comment (Geheiner) [inline sdl-core/src/main/scala/io/smartdatalake/workflow/action/DataFrameActionImpl.scala:347]: what is the difference between evaluating and validating an expectation?

File: sdl-core/src/main/scala/io/smartdatalake/workflow/dataframe/spark/SparkObservation.scala [hunk 14]
```diff
@@ -68,24 +71,50 @@ private[smartdatalake] class SparkObservation(name: String = UUID.randomUUID().t
...
-          notifyAll()
-          sparkSession.foreach(_.listenerManager.unregister(listener))
-        } else {
-          if (logger.isDebugEnabled && qe.observedMetrics.nonEmpty) logger.debug(s"($name) received unexpected metric "+qe.observedMetrics)
-        }
+      val observedMetrics = qe.observedMetrics
+      if (metrics.isEmpty && observedMetrics.isDefinedAt(name)) {
+        logger.debug(s"got observations: ${observedMetrics.keys.mkString(", ")}")
+        metrics = Some(qe.observedMetrics)
+        notifyAll()
```
Comment (Geheiner) [inline sdl-core/src/main/scala/io/smartdatalake/workflow/dataframe/spark/SparkObservation.scala:107]: I'm surprised that we need to do concurrency handling

File: sdl-core/src/main/scala/io/smartdatalake/workflow/dataobject/expectation/Expectation.scala [hunk 15]
```diff
@@ -0,0 +1,236 @@
...
+   * Severity if expectation fails - can be Error (default) or Warn.
+   * If set to Error, execution will fail, otherwise there will be just a warning logged.
+   */
+  def failedSeverity: ExpectationSeverity
+  /**
+   * The aggregation scope used to evaluate the aggregate expression.
+   * Default is 'Job', which evaluates the records transformed by the current job. This is implemented without big performance impacts on Spark.
+   * Other options are 'All' and 'JobPartition', which are implemented by reading the output data again.
+   */
+  def scope: ExpectationScope = Job
```
Comment (Geheiner) [inline sdl-core/src/main/scala/io/smartdatalake/workflow/dataobject/expectation/Expectation.scala:60]: it's not clear for me what difference the scope has for the end result. An example would be nice somewhere, either in a unit test or here: https://smartdatalake.ch/docs/reference/dataQuality

File: sdl-core/src/main/scala/io/smartdatalake/workflow/dataobject/expectation/SQLQueryExpectation.scala [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+import io.smartdatalake.workflow.action.ActionHelper
+import io.smartdatalake.workflow.action.generic.transformer.SQLDfTransformer.INPUT_VIEW_NAME
+import io.smartdatalake.workflow.dataframe.{DataFrameFunctions, GenericColumn, GenericDataFrame}
+import io.smartdatalake.workflow.dataobject.expectation.ExpectationScope.{ExpectationScope, Job}
+import io.smartdatalake.workflow.dataobject.expectation.ExpectationSeverity.ExpectationSeverity
+import io.smartdatalake.workflow.{ActionPipelineContext, DataFrameSubFeed}
+
+
+/**
+ * Definition of an expectation based on a SQL query to be evaluate on dataset-level.
```
Comment (Geheiner) [inline sdl-core/src/main/scala/io/smartdatalake/workflow/dataobject/expectation/SQLQueryExpectation.scala:35]:
```suggestion
 * Definition of an expectation based on a SQL query to be evaluated on dataset-level.
```

File: sdl-core/src/main/scala/io/smartdatalake/workflow/dataobject/expectation/SQLQueryExpectation.scala [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+ * The SQL query will be evaluated in a separate Spark job against the DataFrame.
+ * It supports scope Job and All, but not JobPartition.
+ *
+ * @param code a SQL query returning a single row. All column will be added as metrics.
```
Comment (Geheiner) [inline sdl-core/src/main/scala/io/smartdatalake/workflow/dataobject/expectation/SQLQueryExpectation.scala:39]:
```suggestion
 * @param code a SQL query returning a single row, but potentially multiple columns. All columns will be added as metrics.
```

File: sdl-core/src/main/scala/io/smartdatalake/workflow/dataobject/expectation/UniqueKeyExpectation.scala [hunk 18]
```diff
@@ -0,0 +1,94 @@
...
+      case x: TableDataObject =>
+        assert(x.table.primaryKey.nonEmpty, s"($dataObjectId) 'table.primaryKey' must be defined on DataObject if parameter 'UniqueKeyExpectation.columns' is empty")
+        x.table.primaryKey.getOrElse(throw new IllegalStateException(s"($dataObjectId) 'table.primaryKey' is defined but empty..."))
+      case _ => throw ConfigurationException(s"($dataObjectId) If parameter 'columns' is empty, UniqueKeyExpectation must be defined on a DataObject implementing TableDataObject in order to use primary key")
+    }
+  }
+  override def getAggExpressionColumns(dataObjectId: DataObjectId)(implicit functions: DataFrameFunctions, context: ActionPipelineContext): Seq[GenericColumn] = {
+    val colsToCheck = (if (key.isEmpty) getPrimaryKeyCols(dataObjectId) else key).map(functions.col)
+    Seq(
+      if (scope == ExpectationScope.Job && functions.requestSubFeedType() == typeOf[SparkSubFeed]) Some(functions.approxCountDistinct(functions.struct(colsToCheck:_*), Some(0.005)).as(countDistinctName))
```
Comment (Geheiner) [inline sdl-core/src/main/scala/io/smartdatalake/workflow/dataobject/expectation/UniqueKeyExpectation.scala]: what does magic number 0.005 mean? is that the "high precision"? could be nice to extract it to a variable

File: sdl-core/src/main/scala/io/smartdatalake/util/spark/PushPredicateThroughTolerantCollectMetricsRule.scala [hunk 19]
```diff
@@ -0,0 +1,46 @@
...
+      u match {
+        case m: CollectMetrics if m.name.endsWith(tolerantMetricsMarker) =>
+          u.withNewChildren(Seq(Filter(filter.condition, u.child)))
+        case _ => filter
+      }
+  }
+}
+
+object PushPredicateThroughTolerantCollectMetricsRuleObject {
+  val tolerantMetricsMarker = "!tolerant"
```
Comment (Geheiner) [inline sdl-core/src/main/scala/io/smartdatalake/util/spark/PushPredicateThroughTolerantCollectMetricsRule.scala]: rename to pushdowntolerant ?

File: sdl-core/src/main/scala/io/smartdatalake/workflow/DataFrameSubFeed.scala (hunk 7 above)
Comment (zzeekk) [reply sdl-core/src/main/scala/io/smartdatalake/workflow/DataFrameSubFeed.scala:134]: Type is not Serializable, and this could pop-up.

File: sdl-core/src/main/scala/io/smartdatalake/workflow/action/DataFrameActionImpl.scala (hunk 6 above)
Comment (zzeekk) [reply sdl-core/src/main/scala/io/smartdatalake/workflow/action/DataFrameActionImpl.scala:317]: discussed and ok.

File: sdl-core/src/main/scala/io/smartdatalake/workflow/action/DataFrameActionImpl.scala (hunk 13 above)
Comment (zzeekk) [reply sdl-core/src/main/scala/io/smartdatalake/workflow/action/DataFrameActionImpl.scala:347]: added description

File: sdl-core/src/main/scala/io/smartdatalake/workflow/dataobject/ExpectationValidation.scala (hunk 10 above)
Comment (zzeekk) [reply sdl-core/src/main/scala/io/smartdatalake/workflow/dataobject/ExpectationValidation.scala:75]: Description added

File: sdl-core/src/main/scala/io/smartdatalake/workflow/dataobject/expectation/Expectation.scala (hunk 15 above)
Comment (zzeekk) [reply sdl-core/src/main/scala/io/smartdatalake/workflow/dataobject/expectation/Expectation.scala:60]: Documentation updated

File: sdl-core/src/main/scala/io/smartdatalake/workflow/dataobject/expectation/UniqueKeyExpectation.scala (hunk 18 above)
Comment (zzeekk) [reply sdl-core/src/main/scala/io/smartdatalake/workflow/dataobject/expectation/UniqueKeyExpectation.scala]: Should be calculated from precision -> TODO

**Review by Geheiner — APPROVED** [review]
If you do the 3 TODOs that you agrred are missing then it's fine for me

File: sdl-core/src/main/scala/io/smartdatalake/config/InstanceRegistry.scala (hunk 1 above)
Comment (zzeekk) [reply sdl-core/src/main/scala/io/smartdatalake/config/InstanceRegistry.scala]: Renamed to getDataObjectIdsToValdiateOnRead and adapted implementation.

File: sdl-core/src/main/scala/io/smartdatalake/workflow/action/DataFrameActionImpl.scala (hunk 4 above)
Comment (zzeekk) [reply sdl-core/src/main/scala/io/smartdatalake/workflow/action/DataFrameActionImpl.scala]: You're right, done.

File: sdl-core/src/main/scala/io/smartdatalake/workflow/action/DataFrameActionImpl.scala (hunk 5 above)
Comment (zzeekk) [reply sdl-core/src/main/scala/io/smartdatalake/workflow/action/DataFrameActionImpl.scala]: Ok, done.

File: sdl-core/src/main/scala/io/smartdatalake/workflow/action/DataFrameActionImpl.scala (hunk 9 above)
Comment (zzeekk) [reply sdl-core/src/main/scala/io/smartdatalake/workflow/action/DataFrameActionImpl.scala]: You're right, naming improved.

File: sdl-core/src/main/scala/io/smartdatalake/util/spark/PushPredicateThroughTolerantCollectMetricsRule.scala (hunk 19 above)
Comment (zzeekk) [reply sdl-core/src/main/scala/io/smartdatalake/util/spark/PushPredicateThroughTolerantCollectMetricsRule.scala]: Done

File: sdl-core/src/main/scala/io/smartdatalake/workflow/dataobject/expectation/UniqueKeyExpectation.scala (hunk 18 above)
Comment (zzeekk) [reply sdl-core/src/main/scala/io/smartdatalake/workflow/dataobject/expectation/UniqueKeyExpectation.scala]: Mapping from precision to rsd is not possible as one measure is absolute and the other is relative.
I added configuration options "approximate" and "approximateRsd" to enable approximate distinct calculation.
Default is exact distinct calculation, using a separate query to evaluate.
